Issue #6216: sync content of cli tool in xdoc and cli tool help output

gaurabdg opened the pull request:
Resolves #6216 

- Adds UT for manual sync with both way ignore lists( if we want to ignore an `option` present in xdoc from putting in `Main.java` and vice-versa).

- I added the UT in a new class because `XdocsPagesTest.java` already had a `ClassFanOut` of 25, and adding `picocli.Commandline` made it 26.

- Another UT has been added to check the validity and consistency of the parameters/options present in the `<source>` description.

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:
```diff
@@ -78,60 +78,126 @@
-          + "  -x, --exclude-regexp=<excludeRegex>%n"
-          + "                            Regular expression of directory/file to exclude from"
-          + " CheckStyle%n");
+          "Usage: checkstyle [-dEghjJtTV] [-b=<xpath>] [-c=<configurationFile>] "
```
Note that you change `%n` with `\n` in the test. Please use `%n` as the end-of-line marker to fix windows build.

gaurabdg replied:
@pbludov Done

gaurabdg commented:
@pbludov I mentioned in my first comment why I added the UT in a new file

> * I added the UT in a new class because `XdocsPagesTest.java` already had a `ClassFanOut` of 25, and adding `picocli.Commandline` made it 26.

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/CliOptionsXdocsSyncTest.java:
```diff
@@ -0,0 +1,102 @@
+        final String input = new String(Files.readAllBytes(path), UTF_8);
+
+        final Document document = XmlUtil.getRawXml(path.getFileName().toString(), input, input);
+        final NodeList sections = document.getElementsByTagName("section");
```
Please extract this block to a private static helper method.

gaurabdg replied:
Done

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/CliOptionsXdocsSyncTest.java:
```diff
@@ -0,0 +1,102 @@
+            if (optionsAnnotation != null) {
+                final String option = optionsAnnotation.names()[0];
+                final String descXdoc = cmdDesc.get(option);
+                assertNotNull(descXdoc, "CLI Option" + option + "present in Main.java not "
```
Please use `com.google.common.truth.Truth.assertWithMessage` method for all assertions.

gaurabdg replied:
Done

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/CliOptionsXdocsSyncTest.java:
```diff
@@ -0,0 +1,102 @@
+        final Class<?> cliOptionsClazz = Class.forName("com.puppycrawl.tools.checkstyle"
+                + ".Main$CliOptions");
+        final Field[] fields = cliOptionsClazz.getDeclaredFields();
+        for (Field field:fields) {
```
Please add spaces around `:`

gaurabdg replied:
Done

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/CliOptionsXdocsSyncTest.java:
```diff
@@ -0,0 +1,102 @@
+                assertEquals(descXdoc, optionsAnnotation.description()[0],
+                        "CLI options descriptions should match");
+                cmdDesc.remove(
+                        optionsAnnotation.names()[0]);
```
There is a variable for `optionsAnnotation.names()[0]`

gaurabdg replied:
Done

pbludov commented:
> Is there any workaround?

Of course, the violation can be suppressed, but `ClassFanOut` (and other metrics) is a trigger indicates that the class should be refactored. Extracting this test to its own class is a good idea.

pbludov commented on src/xdocs/cmdline.xml.vm:
```diff
@@ -52,7 +52,7 @@ java -D&lt;property&gt;=&lt;value&gt;  \
       </p>
 
-      <ul>
+      <ul id="CLI Options">
```
Looks like the space is not allowed here

gaurabdg replied:
Yes, I updated. But I wonder why `name="Command line usage"` didn't give an error

gaurabdg replied:
Yes, I updated.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -607,22 +607,31 @@ public boolean isLoggable(LogRecord record) {
 
         /** Config file location. */
-        @Option(names = "-c", description = "Sets the check configuration file to use.")
+        @Option(names = "-c", description = "configurationFile - specifies the location of the"
```
Why did we add `configurationFile` to description?

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -631,14 +640,17 @@ public boolean isLoggable(LogRecord record) {
         @Option(names = {"-g", "--generate-xpath-suppression"},
-                description = "Generates to output a suppression xml to use to suppress all"
-                        + " violations from user's config")
+                description = "--generate-xpath-suppression - generate to output a suppression"
```
same.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -647,33 +659,48 @@ public boolean isLoggable(LogRecord record) {
          */
-        @Option(names = "-f", description = "Sets the output format. Valid values: "
-                + "${COMPLETION-CANDIDATES}. Defaults to ${DEFAULT-VALUE}")
+        @Option(names = "-f", description = "format - specify the output format. Options are"
```
same.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -647,33 +659,48 @@ public boolean isLoggable(LogRecord record) {
         /** Option that controls whether to print the AST of the file. */
         @Option(names = {"-t", "--tree"},
-                description = "Print Abstract Syntax Tree(AST) of the file")
+                description = " --tree - print Abstract Syntax Tree(AST) of the checked file."
```
same and so on.

gaurabdg commented:
@rnveach My bad, did it now. Please review

rnveach commented:
@gaurabdg So they were just added in error from the option's argument name?

gaurabdg commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/CliOptionsXdocsSyncTest.java:
```diff
@@ -0,0 +1,111 @@
+    public void validateCliDocSection() throws Exception {
+        final Map<String, String> cmdDesc = new HashMap<>();
+        final Set<String> ignoreXdocToMain = new HashSet<>();
+        // used to ignore options present in xdocs but not in Main.java
```
Yes, it can be done by creating a mock and querying through given methods

gaurabdg commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -631,14 +640,17 @@ public boolean isLoggable(LogRecord record) {
          * @noinspection CanBeFinal
          */
-        @Option(names = {"-w", "--tabWidth"}, description = "Sets the length of the tab character. "
-                + "Used only with \"-s\" option. Default value is ${DEFAULT-VALUE}")
```
Is, there is. `option.description()` returns the description string by renderning the values.
But there's a strange problem. Some variables like `${COMPLETION-CANDIDATES}` are being rendered correctly but rest `${DEFAULT-VALUE}` is being initialized incorrectly to 0/null.

gaurabdg replied:
It has been resolved, I'll resubmit as soon as I merge the descriptions

gaurabdg commented:
> > Is it checking whether the parameters present in xdocs e.g., --tree, -T, treeWithJavadoc are valid picocli options are not?
> 
> Yes. If we remove `-c` option or add `-xyz` option, we want test to verify that user removed/added options to that display. This issue was started because we are not keeping xdoc up to date with all features.
> 
> For example, https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/cmdline.xml.vm#L37-L45 does not show the option `-e` in it's display.

@rnveach I'll add these changes in this PR and squash them. For now, I have made the description changes. Please review and let me know if some descriptions should change.

pbludov commented:
@gaurabdg please rebase to the latest master, this PR is outdated.

gaurabdg commented:
> > Is it checking whether the parameters present in xdocs e.g., --tree, -T, treeWithJavadoc are valid picocli options are not?
> 
> Yes. If we remove `-c` option or add `-xyz` option, we want test to verify that user removed/added options to that display. This issue was started because we are not keeping xdoc up to date with all features.
> 
> For example, https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/cmdline.xml.vm#L37-L45 does not show the option `-e` in it's display.

@rnveach Done, please review.

romani commented on src/xdocs/cmdline.xml.vm:
```diff
@@ -148,13 +141,21 @@ java -D&lt;property&gt;=&lt;value&gt;  \
+          (must be greater than zero). Default is 1.
+        </li>
+        <li>
+          <code>-W, --tree-walker-threads-number</code> - (experimental) The number of TreeWalker
```
it is sad that we leaked `-C` and `-W` to users, such options do not work.
We should remove them, probably should be done in separate issue.
@rnveach , are you agree ?

gaurabdg replied:
@romani Shall I do it then?

rnveach replied:
I agree. @gaurabdg Please remove them.

gaurabdg replied:
@romani @rnveach I am doing the changes in this issue only, because I revert these changes to the master one, then these options will be present in `Main.java` but not in the xdoc, so I have to add extra lines to ignore these in the tests, which would be redundant once these changes are made in master, and I have to change it again.
Or else if I create a new PR, I have to wait for it to merge and then rebase and push changes here.

rnveach replied:
I assume this PR is blocked by https://github.com/checkstyle/checkstyle/pull/8180

gaurabdg replied:
Yes.

rnveach replied:
This was unblocked and should be rebased if not already.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/CliOptionsXdocsSyncTest.java:
```diff
@@ -0,0 +1,111 @@
+        final String input = new String(Files.readAllBytes(path), UTF_8);
+
+        final Document document = XmlUtil.getRawXml(path.getFileName().toString(), input, input);
+        return document.getElementsByTagName("section");
```
@gaurabdg This is still not done.

gaurabdg replied:
@rnveach `document.getElementById` is returning `null`, since we have to define a separate DTD schema to specify which is actually the ID for this xml. https://stackoverflow.com/a/53044340

rnveach replied:
If we can't get the element by ID, then we don't need the ID in the XDOC.
> `<ul id="CLI_Options">`

gaurabdg replied:
@rnveach It is required for `getListById` method.

rnveach commented on src/xdocs/cmdline.xml.vm:
```diff
@@ -63,76 +67,65 @@ java -D&lt;property&gt;=&lt;value&gt;  \
         <li>
-          <code>-p propertiesFile</code> - specify a
-          properties file to use.
+          <code>-p propertiesFile</code> - Loads the properties file.
```
I feel a combination of how we specify options plus the original description sounds better.
 `Sets the property files to load.`

gaurabdg replied:
@rnveach Done.
